### ISSUE TITLE
Broken pipe when configuration file fails to parse

### ISSUE THREAD

**Issue by kaste** [astral-sh/ruff#9718]
Test scenario:

I have an error in my ruff.toml

```
select = ["F", "E", "W", "B", "ERA", "I001"]
line-length = 100

[lint.isort]
line-length = 90
```

and then I do on Windows CMD

```
type plugin.py | ruff check --no-cache -
```

which outputs

```
> type plugin.py | ruff check --no-cache -
ruff failed
  Cause: Failed to parse C:\Users\c-flo\AppData\Roaming\Sublime Text\Packages\TreeSitter-navigate-around\ruff.toml
  Cause: TOML parse error at line 1, column 1
  |
1 | select = ["F", "E", "W", "B", "ERA", "I001"]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
unknown field `line-length`, expected one of `force-wrap-aliases`, `force-single-line`, `single-line-exclusions`, `combine-as-imports`, `split-on-trailing-comma`, `order-by-type`, `force-sort-within-sections`, `case-sensitive`, `force-to-top`, `known-first-party`, `known-third-party`, `known-local-folder`, `extra-standard-library`, `relative-imports-order`, `required-imports`, `classes`, `constants`, `variables`, `no-lines-before`, `lines-after-imports`, `lines-between-types`, `forced-separate`, `section-order`, `no-sections`, `detect-same-package`, `from-first`, `length-sort`, `length-sort-straight`, `sections`

Ein Prozess hat versucht, zu einer nicht bestehenden Pipe zu schreiben.
```

Note the last line which tells you (in german) that the process tried to write to a closed pipe.  

This is a reduced testcase of course.  I call this within an IDE (Sublime Text, I'm the author of the linting framework SublimeLinter over there) basically like so

```
            try:
                out = proc.communicate(code_b)

            except BrokenPipeError as err:
                print("BrokenPipeError", err)
```
 which prints

```
BrokenPipeError [Errno 32] Broken pipe
```

At this point stdout and stderr are closed so I don't get the actual error message/printing you do on the CLI.  It just aborts here and I can only present the generic "Broken pipe" error.  

Hope this helps.

**Comment by kaste** [comment]
(Also the actual error message is not very indicative as it points to the first line of the conf file which is not the problem, but that's another story)

**Comment by zanieb** [comment]
Hm I also cannot reproduce this in Python

```python
import subprocess

stdin = b"x" * 1000000

process = subprocess.Popen(["ruff", "check", "--no-cache", "-"], stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)

try:
    out = process.communicate(stdin)
except BrokenPipeError as err:
    print(err)
else:
    print(out)

process.wait()
```

```
❯ python example.py
(b'', b'ruff failed\n  Cause: Failed to parse /Users/mz/eng/src/astral-sh/ruff/pyproject.toml\n  Cause: TOML parse error at line 59, column 1\n   |\n59 | [tool.ruff]\n   | ^^^^^^^^^^^\nunknown field `err`\n\n')
``

Do you have a reproducible example?

**Comment by kaste** [comment]
That's indeed surprisingly random.  It seems to depend on the contents of the linted file.  I've set up https://github.com/kaste/ruff-repro  Note that plugin.py results in a Broken Pipe and the very similar plugin_2.py not.  I tried deleting more random lines, the file contents are already fantasy code though without syntax errors, but cannot deduce any pattern here.  Maybe you can reproduce with these files, note however that I'm on Windows.

```
> ruff --version
ruff 0.1.15
```

**Comment by kaste** [comment]
Stupid me, if the file is larger than 4096 it fails.  The standard buffer size on Windows.  🙄

**Comment by kaste** [comment]
I see that you both starred 👀 at my previous message but that's ambiguous.  Do you have an understanding of the bug?  I could then delete the "repro-repo" as it should be clear how to reproduce this.

**Comment by zanieb** [comment]
Could you please provide a reproduction? It's still not clear to me that the broken pipe error is being thrown by Ruff. Generally what happens with broken pipes is

1. Some process is sending data into a pipe
2. Another process (Ruff in this case) is reading data from the pipe
3. The later process encounters an error (invalid configuration in this case) and exits
4. The first process continues sending data to the pipe but the second process has stopped reading it or closed it
5. The pipe is full or closed and is "broken"
6. The first process throws an error

I think it is best practice (and I believe this is commonly implemented) for the first process to _not_ throw the broken pipe error. I don't think Ruff can do anything if the scenario is as-described above.

**Comment by kaste** [comment]
That's a shitty issue here I made.  It is easily resproducable by me but only when
I'm using zombie tools.  Perhaps that's because I'm old too.   🤷 

`cmd.exe` + `type` is of course a real thing in Windows but an ancient one.  Power 
shell's `Get-content` behaves well; and recent versions of python too. To compare 
it with `eslint` is just silly because that's just too slow for any race condition
to appear.

I now backported a sane `Popen._communicate` for my (also ancient) linting framework.
That's 20 lines of code.  

I really was mislead because `type <file> | ruff` was the easy reproduction. 
Actually a combination I never use in practice but only to reproduce or show something.

**Comment by zanieb** [comment]
Sorry I didn't try on Windows I don't have easy access to a machine 😬 

I recently ran into a similar problem piping `git` output into `ripgrep` where the broken pipe error was coming from `git` as I described so I had a hunch this was a problem outside of Ruff. Thanks for taking the time to investigate it — I really appreciate it! I'm glad it's working for you now :)
